connectionmanager/connectionmanager.go

```diff
@@ -0,0 +1,118 @@
+// created the connection manager can accept any number of items supporting initialization,
+// close, or both. The ManageInitialization, ManageClose, and ManageConnection methods can
+// be called as many times as needed in any order to add managed items. They are variadic
+// functions, so multiple items can be added in a single call.
+//
+// Initialization items will immediately start initialization in a separate go routine
+// once the item is added to the ConnectionManager. An internally managed sync.WaitGroup
+// is made available. Calling WG.Wait() on the ConnectionManager will block the current
+// go routine until all initialization functions are complete.
+//
+// Closers are queued up internally running only when the ConnectionManager's Close method
+// is called. The ConnectionManager runs each Close method in a separate go routine and blocks
+// blocks until all are complete.
+type ConnectionManager struct {
+	closerChan chan CloseFunc
+	Close      CloseFunc
+	WG         sync.WaitGroup
+}
+
+// New initializes a new ConnectionManager object that can be used
+// to manage the life of long lived remote connections such as to a database.
+func New(logger *log.Logger) ConnectionManager {
+	closerChan := make(chan CloseFunc)
+	shutdown := make(chan struct{})
+	var stopwg sync.WaitGroup
+	go func() {
+		closers := []CloseFunc{}
+	loop:
+		for {
+			select {
+			case <-shutdown:
+				logger.Println("Shutting Down")
+				break loop
+			case c := <-closerChan:
+				stopwg.Add(1)
+				closers = append(closers, c)
+			}
+		}
+
+		for _, c := range closers {
+			go func(c func()) {
+				c()
+				stopwg.Done()
+			}(c)
+		}
+	}()
+	return ConnectionManager{
+		closerChan: closerChan,
+		Close: func() {
+			shutdown <- struct{}{}
+			stopwg.Wait()
+		},
+	}
+}
+
+// ManageInitialization allows the connection manager to accept any number of items
+// matching the Initializer interface and initializes each in parallel. The wait
+// group is managed to allow callers to block until all managed initialization
+// methods are complete.
+func (cm *ConnectionManager) ManageInitialization(initializers ...Initializer) {
+	for _, initializer := range initializers {
+		cm.WG.Add(1)
+		go func(i Initializer) {
+			i.Initialize()
+			cm.WG.Done()
+		}(initializer)
+	}
+}
+
+// ManageClose allow the connection manager to accept any number of items matching
+// the Closer interface. It queues them up internally. When Close is called on
+// the connection manager, all queued Close methods are executed in parallel.
+// The close method blocks until managed Closers are complete.
+func (cm *ConnectionManager) ManageClose(closers ...Closer) {
+	for _, closer := range closers {
+		cm.closerChan <- closer.Close
+	}
+}
+
+// ManageConnection accepts any number of items matching the InitializerCloser
+// interface and manages both an item's initialization and close.
+func (cm *ConnectionManager) ManageConnection(initializerClosers ...InitializerCloser) {
```

Comment:
ManageConnection seems to restart a connection, close -> initialize -> return. Would a more accurate name be `Restart`?

Reply:
It doesn't...

Though the ordering is a bit confusing, this queues up the close method of the managed item first (which will not run until the connection manager itself has the `Close` method called.  Once queued, it then adds the items initialization function, which immediately starts a go routine that initializes the item. The order ensure the close function is managed before initialization begins.

The fact that you are asking means this is just a bit too confusing. I'll add additional comments to clarify.

Reply:
Ahh I understand, maybe EnqueueConnection/EnqueueClosers/EnqueueInitializers more accurately describes what they're doing?